Support shell less system auditing

igorbeider opened the pull request:
The motivation behind adding shell less auditing is to support system secure auditing in environments where shell command can't be isolated   

- add new Action attribute to check which supports Text and File filtering
- Move FAIL/WARN/PASS constants to common package
- Add new BenchRunner struct that consolidates all relevant methods for CLI and Api usage
- add search by permission for File filtering
- add search by userId and groupId for File filtering

lizrice commented on README.md:
```diff
@@ -1,6 +1,3 @@
-[![Build Status](https://travis-ci.org/aquasecurity/bench-common.svg?branch=master)](https://travis-ci.org/aquasecurity/bench-common)
```
Please leave the build badge there

igorbeider replied:
done

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,334 @@
+	tarHeader tar.Header
+}
+
+func (ctx *tarFileInfo) Name() string {
```
It’s a common convention to use the name `ctx` for a Context - let’s call the receiver something else in all these funcs to avoid possible confusion

igorbeider replied:
done

lizrice commented on actioneval/searchfilterfactory.go:
```diff
@@ -0,0 +1,48 @@
+	Lines   int
+}
+
+type ISearchFilter interface {
```
Conventionally this would be called SearchFilterer

igorbeider replied:
done

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,334 @@
+// Copyright © 2017 Aqua Security Software Ltd. <info@aquasec.com>
```
Copyrights should use current year i.e 2019 for new files, update to 2017-2019 for modifications

igorbeider replied:
done

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,334 @@
+		case common.FilterEntity:
+			filter.filter = val
+		case common.PermissionEntity:
+			if filter.perm, filter.smode, err = parsePermission(val); err != nil {
```
This is more personal preference than a requirement, but i would leave out the if test and the `return nil, err` from all these case statements - you can just have `return filter, err` at the end

igorbeider replied:
done

lizrice commented on actioneval/filesearch_test.go:
```diff
@@ -0,0 +1,304 @@
+}
+
+func TestFileSearchAll(t *testing.T) {
+
```
Have you come across [subtests?](https://blog.golang.org/subtests)? You could avoid lots of repetition in these test cases like that. Tends to make it easier to read what’s being tested, too.

lizrice commented on util/util.go:
```diff
@@ -16,46 +16,41 @@ package util
 
 import (
 	"fmt"
+	"github.com/aquasecurity/bench-common/check"
```
By convention (and by tools like goimports, I think) the github.com ones should be in a block together

igorbeider replied:
done

lizrice commented on common/const.go:
```diff
@@ -0,0 +1,80 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
```
Do we really need a package called `common`?  (Although `utils` is just as egregious to be hones)

lizrice replied:
Most of the contents of this file look like they fit in the actioneval package, except for the test status (maybe that fits better in the check package)

igorbeider replied:
done , common has been removed

lizrice commented on common/errorUtil.go:
```diff
@@ -0,0 +1,19 @@
+	"os"
+)
+
+func HandleError(err error, context string) (errmsg string) {
```
I’ve never liked these two functions, I think it would be better to just inline the printing and/or exiting - I definitely wouldn’t want to use these to justify a package called common

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,334 @@
+	"time"
+)
+
+type tarFileInfo struct {
```
Wondering if it’s really worth having this additional structure, can’t we just use the interfaces on tar.Header directly?

igorbeider replied:
NO
despite original TAR structure implements some of os.FileInfo  methods, 
but return values of these methods do not match the golang FileInfo  file system implementation 
for example the tarHeaders.Mode contains only permission bits while golang os.FileMode  consist of permissions bits and suid/guid bits, hard/soft link/ directory bits etc ... 

so this struct generally used as adapter to os.FileInfo interface

lizrice replied:
Can't you use https://golang.org/pkg/archive/tar/#Header.FileInfo? I think you are pretty much re-implementing everything [here](https://golang.org/src/archive/tar/common.go?s=19247:19286#L519)

As an aside, if you have to have re-implement an interface, it's good practice to have a type assertion to both test and document it (otherwise it's hard to tell why you have implemented several methods).

igorbeider replied:
COOOOL I didn't realized that it exist

igorbeider replied:
done

lizrice commented on cmd/root.go:
```diff
@@ -1,13 +1,26 @@
 	homedir "github.com/mitchellh/go-homedir"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"os"
```
A few examples of this where it looks like your editor has done strange things to import ordering

igorbeider replied:
I created this structure as adapter to os.FileInfo interface 
later the code logic iterates files files from file system or tar headers in similar manner using os.FileInfo.

igorbeider replied:
done

lizrice reviewed (commented):
I like a LOT of the restructuring here. 

I haven’t had time to look at the new functionality properly yet, but I did find a few Go convention things that it would be better to sort out.

igorbeider commented:
@lizrice 
Thank you Liz for your Input , I will address your comments as soon as I am finishing my duties. 
Another Q, since the change might break exiting projects that relies on bench-common I would like  to adjust it to use newer bench-common version, could you please point me out what projects are using this library ?

lizrice commented:
I believe it's docker-bench and linux-bench. I don't think kube-bench uses it (yet). @yoavAqua have I missed anything?

yoavAqua commented:
You are correct.
I do think that we should change kube-bench to use the common engine.
But it's not the time I guess.


From: Liz Rice
Sent: Friday, May 3, 16:23
Subject: Re: [aquasecurity/bench-common] Support shell less system auditing  (#31)
To: aquasecurity/bench-common
Cc: Yoav Hizkiahou, Mention


I believe it's docker-bench and linux-bench. I don't think kube-bench uses it (yet). @yoavAqua<https://github.com/yoavAqua> have I missed anything?
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/aquasecurity/bench-common/pull/31#issuecomment-489093373>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKV26N7B6EIT5YPB5LXLYNTPTQ4DBANCNFSM4HHVNGUQ>.

igorbeider commented:
Hi Liz , 
I have addressed some of your comments and made the necessary fix. 
1) Please see my comment regarding tar headers
2) regarding common package, generally I introduced it to keep there the HandleError method and 
and constants PASS/FAIL/WARN/INFO , because it used widely in all other packages.
Leaving it in check package creates "import cycle" issues

lizrice commented on actioneval/textsearch_test.go:
```diff
@@ -0,0 +1,157 @@
+)
+
+//create tmp file for that will hold test content
+var tmpFile = fmt.Sprintf("%v-%d", "/tmp/test_text_search_content", os.Getpid())
```
Any reason not to use ioutil.TempFile()?

igorbeider replied:
done

lizrice commented on util/util_test.go:
```diff
@@ -15,11 +15,15 @@
 	"testing"
 )
 
+var tmpFile = fmt.Sprintf("%v,%d", "/tmp/test_util_text_search_content", os.Getpid())
```
As before this could be `ioutil.TempFile()` (if this is used somewhere?)

igorbeider replied:
done

igorbeider commented:
@lizrice @yoavAqua 
I have addressed all comments 
However tried to update linux-bench & docker-bench, but failed due to missing permission to upstream my branch

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+	if info.Sys() != nil {
+
+		if sys, ok := info.Sys().(*syscall.Stat_t); ok {
+			uid = uint32(sys.Uid)
```
I think `sys.Uid` and `sys.Gid` are already uint32 and don't need casting

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+		}
+	}
+
+	if f.userId != -1 && uint32(f.userId) != uid {
```
And you could define `f.userId` and `f.groupId` as uint32 so you don't need to cast them either. (It's a good idea to avoid casting if you can, to avoid the possibility of losing information.) 

And then, this code could be simplified by not having the local variables `uid` and `gid` and simply comparing directly against `sys.Uid` and `sys.Gid`

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+
+	tarHeaders []tar.Header
+
+	groupId int64
```
You should define these as uint32 to avoid casting later on, and I think you could have 0 represent "not defined"

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+	return result
+}
+
+func (f *FileSearchFilter) satisfyGroupIdAnUserIdFilter(info os.FileInfo) bool {
```
Nit: would be nice to correct the spelling to satisfyGroupIdAn**d**UserIdFilter

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+		if permInt, err := strconv.ParseInt(reg.ReplaceAllString(perm, ""), 8, 64); err != nil {
+			return 0, 0, err
+		} else {
+			return int64(convertMode(permInt)), mode, nil
```
Where the type is under your control, define it to be what you want to avoid unnecessary casting

lizrice commented on runner/BenchRunner.go:
```diff
@@ -0,0 +1,186 @@
+
+func New(configYaml []byte) (runner *BenchRunner) {
+
+	ctx := new(BenchRunner)
```
Please only use `ctx` for actual Go[ contexts](https://golang.org/pkg/context/)

lizrice commented on runner/BenchRunner.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func (ctx *BenchRunner) createControls() (err error) {
+	ctx.mControls, err = check.NewControls(ctx.mConfigYaml).WithIsAction(ctx.mIsActionTest).
```
This sort of functional chaining is quite fun, and I suppose it works so I'm not going to make you remove it, but it feels kinda overkill to me given that all of the "WithX" functions just assign field values. Personally I'd have kept it simple and just assigned the fields directly into the struct. "Clear is better than clever" as the [proverb ](https://go-proverbs.github.io/) goes.

lizrice commented on util/util.go:
```diff
@@ -69,6 +62,9 @@ func continueWithError(err error, msg string) string {
 func CleanIDs(list string) []string {
+	if list == "" {
+		return nil
+	}
```
Is this necessary? I think the subsequent code will handle it regardless (and less code is easier to read)

lizrice commented on check/controls.go:
```diff
@@ -40,60 +47,93 @@ type Summary struct {
 	Info int `json:"total_info"`
 }
 
+func (controls *Controls) WithIsAction(isAction bool) *Controls {
```
See below my comment about functional chaining... it's OK, but it's a heavyweight way to assign some fields!

Commit:
This commit adds support for New type of attribute called Action

The action attributes provide support for TextSearch and FileSearch
functionality over predefined file system.

We also made some minor refactor to the tool to support more convenient way for
API usage, now not need to copy/paste part of 'becnh' code, instead use
runner.BenchRunner struct that contains all necessary API.

Tar headers support has been implemented to address FileSearch over external tar files.
Now whatever are using this lib, can inject tar headers instead real FS
and get desired results for FileSearch actions

Add more unit test for file search action
Resolve conflicts

igorbeider commented:
@lizrice  
HI Liz , 
I have made several changes today to this branch 
1) re-base vs master 
2) added more unit tests 
3) addressed some of your comments regarding unit testing.

regarding all the With* methods , we have been using it a lot in our production code , it also called 'builder patter' 
if you insist ,I will remove it, just let me know.

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+	if len(perm) == 4 || len(perm) == 3 {
+		if permInt, err := strconv.ParseInt(reg.ReplaceAllString(perm, ""), 8, 64); err != nil {
+			return 0, 0, err
+		} else {
```
Convention is not to use an else branch if you return from the if branch

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+
+	perm = reg.ReplaceAllString(perm, "")
+	if len(perm) == 4 || len(perm) == 3 {
+		if permInt, err := strconv.ParseInt(reg.ReplaceAllString(perm, ""), 8, 64); err != nil {
```
In line 278 you've removed all the non-numeric chars, so why do it again in line 280?

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+		mode = util.ModeExact
+	}
+	//strip non numeric chars
+	reg, err := regexp.Compile("\\D")
```
Would be better to just compile the regex once (could use MustCompile). I can't help wondering if there isn't a better way to extract the numeric chars that you want here.

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+// first three bits represents the sticky bit/setuid/setgid
+// other 9 bits represents the file mode i.e 777
+// in order to compare it with os.FileMode, need to convert to 32 bits
+func convertMode(mode int64) uint64 {
```
Define mode as uint64 if that's what you need it to be

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+	}
+
+	// check if we satisfy the permission filter condition
+	if f.perm != 0 && f.sMode != 0 && !f.satisfyPermissionFilter(info) {
```
why not do the tests on perm and sMode inside satisfyPermissionFilter(), it would be easier to read

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+func (f *FileSearchFilter) satisfyAllFilters(info os.FileInfo) bool {
+
+	if f.filter != "" &&
+		!f.satisfyFilter(info.Name()) {
```
I'd do the test on f.filter inside satisfyFilter for clarity

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+		return false
+	}
+
+	if f.groupId != -1 && uint32(f.groupId) != gid {
```
Casting

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+
+	// ensure that search location does not escape the workspace
+	if !strings.HasPrefix(clearRootPath, workspacePath) {
+		result.Errmsgs += util.HandleError(fmt.Errorf("relative path "+rootPath+" is not supported "), reflect.TypeOf(f).String())
```
no need for += here, result is a new variable

lizrice commented on actioneval/filesearch.go:
```diff
@@ -0,0 +1,296 @@
+	filter.userId = -1
+	filter.groupId = -1
+	for _, mapItem := range mapSlice {
+		key, val := parseMapKeyValToString(mapItem)
```
I haven't thought about this very deeply but doesn't the yaml package make it easier to parse the content than this? I'd expect you could just unmarshal it?

lizrice commented:
I've reviewed a couple more functions in detail and found some more issues. There is quite a lot I haven't really looked at properly yet.

The problem here is that this is a GIANT PR, so it's hard to review - it would be much easier to cope with if it were done in digestible chunks with one change at a time. 

It would also be helpful to understand why these changes are needed. I don't have any context about the kinds of tests that need these changes. Do they look like *-bench YAML config files, like the ones we use for kube-bench / docker-bench / linux-bench? This would probably be easier if we had a conversation about it rather than trying to do it here in GitHub :-)
